### DISCUSSION TITLE
STM32H7XX additional RAM

### DISCUSSION THREAD

**Discussion by saraverbeecke** [micropython/micropython#11582] [STM32 / Pyboard] [answered]
Hi, I'm planning to develop a custom STM32H74X board but I want to extend the 1024kB of RAM. Anyone experience with this to use with micropython?
I've seen the Portenta H7 port, but it has too many features. I want to create a port just fot the STM32H7XX and access the external SDRAM. Maybe other simpler STM32H7 boards exist with extra external RAM, but did not find any...

**Comment by brave-ulysses** [comment]
aps6404 chip comes up on h723 nucleo board. haven't tried hw spi or actually using memory for anything...

```
print(f'\n\nstarting spi ram')
# psram hw [q]spi1 interface. cs=b6 io1/miso=d12 clk=b2 io0/mosi=d11 io3/rst=d13 io2/wp=e2
from machine import SPI, Pin
from spiram import SPIRAM
spi_psram_h = machine.SoftSPI(miso=machine.Pin.cpu.D12,mosi=machine.Pin.cpu.D11,sck=machine.Pin.cpu.B2,baudrate=1000000,bits=8,polarity=0,phase=1)
cspins = ( (Pin(Pin.cpu.B6, Pin.OUT, value=1)), )
# cs_psram_h  = machine.Pin(machine.Pin.cpu.B6, mode=machine.Pin.OUT, value=1)
ram = SPIRAM(spi_psram_h, cspins)
print(f'completing spi ram\n\n')
```

```
>>> from machine import SPI, Pin
>>> from spiram import SPIRAM
>>> spi_psram_h = machine.SoftSPI(miso=machine.Pin.cpu.D12,mosi=machine.Pin.cpu.D11,sck=machine.Pin.cpu.B2,baudrate=1000000,bits=8,polarity=0,phase=1)
>>> cspins = ( (Pin(Pin.cpu.B6, Pin.OUT, value=1)), )
>>> ram = SPIRAM(spi_psram_h, cspins)
Total SPIRAM size 8192 KiB in 1 devices.
>>> 
```

**Comment by jimmo** [comment, marked as answer]
> I've seen the Portenta H7 port, but it has too many features.

You just want the SDRAM config from mpconfigport.h. For a simpler example look at `STM32F429DISC/mpconfigport.h`. The point is that you define all the `MICROPY_HW_SDRAM_*` macros to configure it, and then override `HEAP_START` and `HEAP_END` so that the SDRAM gets used for the heap. Everything else happens automaticaly (main.c will call sdram_init() and sdram.c implements the driver).

**Reply by saraverbeecke** [reply]
@jimmo thanks for pointing me there to setup the right configuration variables. 
But I first need to add the RAM chip itself :-)... I am not an expert in RAM chips. I have a STM32H723 sample here, do you know if I should go for SRAM, SDRAM or PSRAM using the FMC or what about the OCTOSPI memory-interface... 
Do you have RAM chip suggestions (I'm looking for at least 64 MBit)?

**Reply by 4riceotter** [reply]
@saraverbeecke The answer to the Which RAM question is always “depends on what you need” ;-), but if you have to ask it here, then you’d want the SDRAM :-).  SRAM is more expensive per bit, while DDR has low cost per bit, but requires a lot of work (HW & FW) to support, and I’m not sure if the STM32 can support it.  You can find SDRAM implementation example on all the STM32 discovery cards with schematic & code, so that’s what I’d recommend.
Benjamin

**Reply by saraverbeecke** [reply]
Thanks @4riceotter
